Update MUI from v4 to v5

czmj opened the pull request:
This upgrade is done by following the [official document](https://mui.com/material-ui/migration/migration-v4) on the upgrade. This document is "getting started" on doing MUI upgrade, but we don't need to make (manually) any changes listed in other migration documents, because they are done by a migration tool from MUI called codemods.

In the migration guide, we can see that we meet the requirements for doing the upgrade, like minimal Typescript, React and Node.js version.

We can proceed with installation of new packages; `@mui/material` and `@mui/styles` are needed to replace old `@material-ui/core`, which is now removed.

> One of the biggest changes in v5 is the replacement of JSS for [Emotion](https://emotion.sh/docs/introduction) as a default styling solution.

However, we were using `styled-components` already and will continue doing so. Therefore, instead of installing `@emotion/react` and `@emotion/styled` peer dependencies, we install `@mui/styled-engine-sc`. We also need to [point Webpack](https://mui.com/material-ui/guides/styled-engine/#how-to-switch-to-styled-components) to use this package as the default styled engine.

Imports are changed and full list of changes can be found [here](https://mui.com/material-ui/migration/migration-v4/#replace-all-imports). But, this part is automatically migrated by MUI codemods:
```
npx @mui/codemod v5.0.0/preset-safe apps
npx @mui/codemod v5.0.0/preset-safe libs
```

Lastly, after upgrading to MUI v5, wrapping `ThemeProvider` from `@mui/material` inside another `ThemeProvider`, but from `styled-components` seems to swallow outer `theme` props, so we include all theme props inside MUI `ThemeProvider` and reduce the wrapping.

Seems like the rest of breaking changes aren't affecting us.

czmj committed:
Change MUI element default margins

MUI has changed the default margin from
"normal" to "none" on form components.

CDM doesn't actually use any MUI form
components so it's not a huge problem,
but we're setting the default to what
we will want for FTS Admin.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 committed:
Change MUI imports after migrating to v5

With the release of v5, the names of all related packages
were changed from @material-ui/* to @mui/* as part of MUI
updated branding.
https://mui.com/material-ui/migration/migration-v4/#replace-all-imports

This commit is the partial output (some changes are omitted)
of running MUI codemods for automated migration:
```
npx @mui/codemod v5.0.0/preset-safe apps
npx @mui/codemod v5.0.0/preset-safe libs
```

Co-authored-by: Clara Jordan <clara.jordan@hpc.tools>

Pl217 committed:
HPC-8326: Upgrade to MUI v5

One of the biggest changes in v5 is the replacement
of JSS for [Emotion](https://emotion.sh/docs/introduction)
as a default styling solution.

Since we already have been using `styled-components`
we need to install `@mui/styled-engine-sc` to replace
the default Emotion styled engine.

Co-authored-by: Clara Jordan <clara.jordan@hpc.tools>

Pl217 committed:
Use `styled-components` as MUI styled engine

Pl217 committed:
Configure `styled-components` as MUI styled engine in TS config

Pl217 committed:
Provide all theming via MUI ThemeProvider

For some reason, after upgrading to MUI v5,
theme properties from `THEME` variable are
not readable as theme props.

If element's style is defined with `styled`,
for example:
```
styled.div`
  margin-top: ${(p) => p.theme.marginPx.lg}px;
`
```
`p.theme` would not have access to `marginPx`

Pl217 commented on package.json:
```diff
@@ -44,7 +44,8 @@
     "@babel/preset-react": "7.16.7",
     "@babel/preset-typescript": "7.16.7",
-    "@material-ui/core": "^4.12.3",
+    "@emotion/react": "^11.9.0",
+    "@emotion/styled": "^11.8.1",
```
In the [upgrade guide](https://mui.com/material-ui/migration/migration-v4/#introduction), inside introductory section, they say:

> One of the biggest changes in v5 is the replacement of JSS for [Emotion](https://emotion.sh/docs/introduction) as a default styling solution.

But, we are already using `styled-components`. In the [peer dependency section](https://mui.com/material-ui/migration/migration-v4/#peer-dependencies), they say you should use `@mui/styled-engine-sc` instead of Emotion packages.

I see you installed Emotion packages here (devDependency) and `@mui/styled-engine-sc` as a regular dependency.

Pl217 commented on package.json:
```diff
@@ -105,8 +106,13 @@
     "ts-jest": "27.1.4",
     "typescript": "4.6.3"
+  },
+  "dependencies": {
+    "@mui/material": "^5.8.1",
+    "@mui/styled-engine-sc": "^5.8.0",
+    "@mui/styles": "^5.8.0",
+    "styled-components": "^5.3.5"
```
Why did you move the packages from `devDependencies` to regular dependencies? I don't know why, but all packages were marked as `devDependencies`. Maybe that's a Nx thing, but let's keep it for now.

Pl217 commented on libs/hpc-ui/src/lib/theme.tsx:
```diff
@@ -82,6 +82,18 @@ export const THEME = {
 } as const;
 
 export const MUI_THEME = createTheme({
+  components: {
+    MuiTextField: {
+      defaultProps: {
+        margin: 'normal',
+      },
+    },
+    MuiFormControl: {
+      defaultProps: {
+        margin: 'normal',
+      },
+    },
+  },
```
Why are these changes made? When I run [codemods](https://mui.com/material-ui/migration/migration-v4/#run-codemods), these changes aren't made.

czmj replied:
Ok, thanks for your changes. Please merge them in to this branch.

The changes to margins are there because MUI has changed the default margin from "normal" to "none" on form components. CDM doesn't actually use any MUI form components so it's not a huge problem, but setting the default to what we will want for FTS Admin now made sense to me - I can move it to a separate PR if you'd like.

Pl217 reviewed (changes requested):
My main complaint is that you haven't documented anything regarding this upgrade. I know this is extracted from #305, but we cannot have a single commit saying "Update MUI from v4 to v5" and call it a day.

I prefer git commits split between individual actions you took while doing the update. This may require you to skip commit checks (or make git bisecting harder later), but makes the upgrade process much easier to follow compared to a single big commit.

Pointing to the document used while doing the upgrade, commenting on specific parts, etc. all of that is lacking on this PR. I have gone through the update process myself and addressed my main complaints from code comments.

`styled-components` are used instead of Emotion, which means `@mui/styled-engine-sc` needed to be setup with Webpack and Typescript.
Also, I stumbled upon a problem where wrapping `ThemeProvider` from `@mui/material` inside another `ThemeProvider`, but from `styled-components` (this is all inside `libs/hpc-ui/src/lib/theme.tsx` file) seems to swallow outer `theme` props, so we need to include all theme props inside MUI `ThemeProvider` and reduce the wrapping.

This is all pushed as [pp/HPC-8326](https://github.com/UN-OCHA/hpc-cdm/tree/pp/HPC-8326) branch, so you can check it out. I will also update this PRs first comment with update process I used. Please review my commits and if you agree with the changes, I will move them over to this branch.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
